latch/registry/table.py

```diff
@@ -0,0 +1,275 @@
+                        typ = column["type"]
+
+                        for data_point in record_data:
+                            if data_point["key"] != key:
+                                continue
+                            values[key] = to_python_literal(
+                                data_point["data"],
+                                typ["type"],
+                            )
+                            break
+
+                        if key not in values:
+                            if not typ["allowEmpty"]:
+                                raise ValueError(
+                                    f"no value provided for required column {key}"
+                                )
+                            values[key] = None
+                except ValueError as e:
+                    errors.append(
+                        ValueError(
+                            f"record {record['id']} ({record['name']}) is invalid: {e}"
+                        )
+                    )
+                    continue
+
+                output.append(
+                    Record(
+                        id=record["id"],
+                        name=record["name"],
+                        _value_dict=values,
+                    )
+                )
+
+            yield {"records": output, "errors": errors}
+
+    def update(self):
+        return TableUpdater(self)
+
+
+@dataclass
+class TableUpdate:
+    table: Table
+
+    async def execute(self, session: gql.Client):
+        ...
+
+
+# todo(ayush): "DeleteRecordUpdate", "UpsertColumnUpdate", "DeleteColumnUpdate"
+@dataclass
+class UpsertRecordUpdate(TableUpdate):
+    name: str
+    data: Dict[str, Any]
+
+    async def execute(self, session: gql.Client):
+        record_id = await session.execute(
+            gql.gql(
+                """
+                mutation CatalogMultiUpsertSamples($tableId: BigInt!, $names: [String!]!) {
+                    catalogMultiUpsertSamples(
+                        input: { argExperimentId: $tableId, argNames: $names }
+                    ) {
+                        bigInts
+                        clientMutationId
+                    }
+                }
+                """
+            ),
+            {
+                "tableId": self.table.id,
+                "names": [self.name],
+            },
+        )["catalogMultiUpsertSamples"]["bigInts"][0]
+
+        for column in self.table.get_columns():
+            literal = None
+            if column["key"] not in self.data:
+                if not column["type"]["allowEmpty"]:
+                    raise ValueError(
+                        f"unable to upsert record {self.name} as column"
+                        f" {column['key']} is missing a value"
+                    )
+            else:
+                literal = to_registry_literal(
+                    self.data[column["key"]],
+                    column["type"]["type"],
+                )
+
+            await session.execute(
```

Comment:
these very much need to be all in the same query and also probably the same function as the upsert thing

Reply:
at least leave a `todo`